common/cryptfs.c

```diff
@@ -622,97 +763,82 @@ cryptfs_setup_volume_integrity_new(const char *label, const char *real_blkdev,
+			DEBUG("Successfully formatted volume %s using file_copy", crypto_blkdev);
 		}
 	}
-	return crypto_blkdev;
-error:
-	mem_free0(crypto_blkdev);
-	mem_free0(integrity_dev);
-	return NULL;
-}
 
-char *
-cryptfs_setup_volume_new(const char *label, const char *real_blkdev, const char *key,
-			 const char *meta_blkdev)
-{
-	int fd;
-	// The file system size in sectors
-	uint64_t fs_size;
-
-	/* Update the fs_size field to be the size of the volume */
-	if ((fd = open(real_blkdev, O_RDONLY)) < 0) {
-		ERROR("Cannot open volume %s", real_blkdev);
-		return NULL;
+	if (crypto_key) {
+		mem_memset0(crypto_key, crypto_key_len);
+		mem_free0(crypto_key);
 	}
-	// BLKGETSIZE64 returns size in bytes, we require size in sectors
-	int sector_size = dm_get_blkdev_sector_size(fd);
-	if (!(sector_size > 0)) {
-		ERROR("dm_get_blkdev_sector_size returned %d\n", sector_size);
-		return NULL;
-	}
-	fs_size = dm_get_blkdev_size64(fd) / sector_size;
-	close(fd);
-
-	if (fs_size == 0) {
-		ERROR("Cannot get size of volume %s", real_blkdev);
-		return NULL;
+	if (integrity_key) {
+		mem_memset0(integrity_key, integrity_key_len);
+		mem_free0(integrity_key);
 	}
-	DEBUG("Crypto blk device size: %" PRIu64, fs_size);
-
-	if (meta_blkdev)
-		return cryptfs_setup_volume_integrity_new(label, real_blkdev, meta_blkdev, key,
-							  fs_size);
-
-	// do dmcrypt device setup only
+	if (integrity_dev_label)
+		mem_free0(integrity_dev_label);
+	if (integrity_blkdev && integrity_blkdev != crypto_blkdev)
+		mem_free0(integrity_blkdev);
 
-	/* Use only the first 64 hex digits of master key for 512 bit xts mode */
-	IF_TRUE_RETVAL(strlen(key) < CRYPTFS_FDE_KEY_LEN, NULL);
-	char enc_key[CRYPTFS_FDE_KEY_LEN + 1];
-	memcpy(enc_key, key, CRYPTFS_FDE_KEY_LEN);
-	enc_key[CRYPTFS_FDE_KEY_LEN] = '\0';
+	return crypto_blkdev;
 
-	if (create_crypto_blk_dev(real_blkdev, enc_key, label, fs_size, false) < 0)
-		return NULL;
+error:
+	if (crypto_key) {
+		mem_memset0(crypto_key, crypto_key_len);
+		mem_free0(crypto_key);
+	}
+	if (integrity_key) {
+		mem_memset0(integrity_key, integrity_key_len);
+		mem_free0(integrity_key);
+	}
+	if (integrity_dev_label)
+		mem_free0(integrity_dev_label);
+	if (integrity_blkdev) {
+		delete_integrity_blk_dev(label);
+		mem_free0(integrity_blkdev);
+	}
+	if (crypto_blkdev) {
+		delete_crypto_blk_dev(-1, label);
+		mem_free0(crypto_blkdev);
+	}
 
-	return create_device_node(label);
+	return NULL;
 }
 
 int
-cryptfs_delete_blk_dev(int fd, const char *name)
+cryptfs_delete_blk_dev(int fd, const char *name, cryptfs_mode_t mode)
 {
-	char buffer[DM_CRYPT_BUF_SIZE];
-	struct dm_ioctl *io;
-	int ret = -1;
-
-	io = (struct dm_ioctl *)buffer;
-
-	dm_ioctl_init(io, INDEX_DM_DEV_REMOVE, DM_CRYPT_BUF_SIZE, name, NULL, 0, 0, 0, 0);
-	if (dm_ioctl(fd, DM_DEV_REMOVE, io) < 0) {
-		ret = errno;
-		if (errno != ENXIO)
-			ERROR_ERRNO("Cannot remove dm-crypt device");
-		goto error;
+	bool encrypt = true, integrity = false;
```

Comment:
i'd suggest setting both values in each switch-case branch and skip setting them here